### PR TITLE
oauth token caching

### PR DESCRIPTION
**Pull request by NRHelmi**
This PR adds support to oauth token caching

### CODE REVIEW

File: railib/rest.py [hunk 1]
```diff
@@ -119,13 +121,52 @@ def _print_request(req: Request, level=0):
...
+    return path.join(path.expanduser('~'), '.rai', 'tokens.json')
+
+
+# Read oauth cache
+def _read_cache() -> dict:
+    try:
+        with open(_cache_file(), 'r') as cache:
+            return json.loads(cache.read())
+    except Exception:
+        return None
```
Comment (geokollias) [inline railib/rest.py:134]: I see we swallow these exceptions in these cache functions (the go sdk does sth similar as well). Is that on purpose?

File: railib/rest.py [hunk 2]
```diff
@@ -119,13 +121,52 @@ def _print_request(req: Request, level=0):
...
+# write access token to cache
+def _write_token_cache(creds: ClientCredentials):
+    try:
+        cache = _read_cache()
+        if cache:
+            cache[creds.client_id] = creds.access_token
+        else:
+            cache = {creds.client_id: creds.access_token}
+        with open(_cache_file(), 'w') as f:
+            f.write(json.dumps(cache, default=vars))
```
Comment (geokollias) [inline railib/rest.py:155]: what's `default=vars` needed for?

File: railib/rest.py [hunk 3]
```diff
@@ -119,13 +121,52 @@ def _print_request(req: Request, level=0):
...
+
+
+# write access token to cache
+def _write_token_cache(creds: ClientCredentials):
+    try:
+        cache = _read_cache()
+        if cache:
+            cache[creds.client_id] = creds.access_token
+        else:
+            cache = {creds.client_id: creds.access_token}
```
Comment (geokollias) [inline railib/rest.py:153]: maybe we could avoid this branch by returning `{}` from `_read_cache`'s exception path

File: railib/rest.py [hunk 4]
```diff
@@ -119,13 +121,52 @@ def _print_request(req: Request, level=0):
...
+            return json.loads(cache.read())
+    except Exception:
+        return None
+
+
+# Read access token from cache
+def _read_token_cache(creds: ClientCredentials) -> AccessToken:
+    try:
+        cache = _read_cache()
+        return AccessToken(**cache[creds.client_id])
```
Comment (geokollias) [inline railib/rest.py:141]: I assume if the cache doesn't contain that client id an exception is thrown?

**Review by geokollias — APPROVED** [review]
LGTM Helmi! Have added some (mostly minor) comments/questions.

File: railib/rest.py (hunk 1 above)
Comment (NRHelmi) [reply railib/rest.py:134]: as far as I do understand, if the cache is not accessible for some reason we should not block the user and instead we request a new token

File: railib/rest.py (hunk 2 above)
Comment (NRHelmi) [reply railib/rest.py:155]: this is one way to make python class json serializable

File: railib/rest.py (hunk 4 above)
Comment (NRHelmi) [reply railib/rest.py:141]: yep that's true and we just by-pass it, if `_read_token_cache` function return `None` we will request a new token anyway

**Pull request merged**